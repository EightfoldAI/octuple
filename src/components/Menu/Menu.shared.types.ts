import { MenuItemTypes, MenuSize, MenuVariant } from './Menu.types';
import { ReactNode } from 'react';

export interface MenuRendererProps {
  items: MenuItemTypes[];
  size?: MenuSize;
  variant?: MenuVariant;
  children?: ReactNode;
  [key: string]: any;
}

export interface MenuRenderFunction {
  (props: MenuRendererProps): JSX.Element;
}
