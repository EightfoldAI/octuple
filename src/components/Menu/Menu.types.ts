import React from 'react';
import { ListProps } from '../List';
import {
  MenuItemButtonProps,
  MenuItemCustomProps,
  MenuItemLinkProps,
  MenuItemSubHeaderProps,
  DropdownMenuItemProps,
} from './MenuItem/MenuItem.types';
import { ButtonProps } from '../Button';

export enum MenuVariant {
  disruptive = 'disruptive',
  primary = 'primary',
  neutral = 'neutral',
}

export enum MenuSize {
  large = 'large',
  medium = 'medium',
  small = 'small',
}

export type MenuItemTypes =
  | MenuItemLinkProps
  | MenuItemButtonProps
  | MenuItemSubHeaderProps
  | MenuItemCustomProps
  | DropdownMenuItemProps;

export interface MenuProps
  extends Omit<
    ListProps<MenuItemTypes>,
    'renderItem' | 'role' | 'itemRole' | 'footer'
  > {
  /**
   * Props for the cancel button
   */
  cancelButtonProps?: ButtonProps;
  /**
   * Props for the ok button
   */
  okButtonProps?: ButtonProps;
  /**
   * Callback when cancel button is clicked
   */
  onCancel?: React.MouseEventHandler<HTMLButtonElement>;
  /**
   * On change callback when menu item is clicked
   * @param value
   */
  onChange?: (value: any) => void;
  /**
   * Callback when ok button is clicked
   */
  onOk?: React.MouseEventHandler<HTMLButtonElement>;
  /**
   * The menu aria role.
   * @default 'menu'
   */
  role?: string;
  /**
   * Size of the menu
   * @default MenuSize.Medium
   */
  size?: MenuSize;
  /**
   * SubHeader of the menu
   * @deprecated 'Unused prop'
   */
  subHeader?: string;
  /**
   * Variant of the menu item
   * @default MenuVariant.neutral
   */
  variant?: MenuVariant;
}

export interface DropdownMenuProps extends MenuProps {
  /**
   * If the dropdown is disabled or not
   */
  disabled?: boolean;
  /**
   * Callback executed on reference element click.
   * @param event
   * @returns (event: React.MouseEvent) => void
   */
  referenceOnClick?: (event: React.MouseEvent) => void;
  /**
   * Callback executed on reference element keydown.
   * @param event
   * @returns (event: React.KeyboardEvent) => void
   */
  referenceOnKeydown?: (event: React.KeyboardEvent) => void;
}
