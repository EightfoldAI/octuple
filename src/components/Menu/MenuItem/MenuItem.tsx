import React, {
  FC,
  ForwardedRef,
  forwardRef,
  JSXElementConstructor,
} from 'react';
import { MenuItemProps, MenuItemType } from './MenuItem.types';
import { MenuItemButton } from './MenuItemButton';
import { MenuItemCustom } from './MenuItemCustom';
import { MenuItemLink } from './MenuItemLink';
import { MenuItemSubHeader } from './MenuItemSubHeader';

const TYPE_TO_MENU_ITEM_MAP: Record<
  MenuItemType,
  JSXElementConstructor<any>
> = {
  [MenuItemType.button]: MenuItemButton,
  [MenuItemType.custom]: MenuItemCustom,
  [MenuItemType.link]: MenuItemLink,
  [MenuItemType.subHeader]: MenuItemSubHeader,
};

export const MenuItem: FC<MenuItemProps> = forwardRef(
  ({ type, ...rest }, ref: ForwardedRef<any>) => {
    const Component = TYPE_TO_MENU_ITEM_MAP?.[type] ?? null;
    return <Component ref={ref} {...rest} />;
  }
);
