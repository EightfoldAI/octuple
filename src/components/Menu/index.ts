export * from './Menu';
export * from './Menu.types';
export * from './MenuItem/MenuItem.types';
