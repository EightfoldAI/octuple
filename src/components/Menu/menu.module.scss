.menu-container {
  background: var(--menu-background-color);
  display: flex;
  flex-direction: column;

  .menu-header-container {
    color: var(--menu-header-color);
    display: flex;
    flex: 1;

    &.large {
      @include octuple-h4();
      padding: $space-ml $space-ml $space-xs;
    }

    &.medium {
      @include octuple-h5();
      padding: $space-m $space-m $space-xs;
    }

    &.small {
      @include octuple-h6();
      padding: $space-s $space-s $space-xs;
    }

    .heading {
      flex: 1;
    }
  }

  .menu-footer-container {
    background: var(--menu-footer-background-color);
    position: sticky;
    bottom: 0;
    border-top: solid 1px var(--menu-border-color);

    &.large {
      padding: $space-l;
    }

    &.medium {
      padding: $space-m;
    }

    &.small {
      padding: $space-s;
    }
  }
}
