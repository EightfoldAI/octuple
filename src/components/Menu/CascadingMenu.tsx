'use client';

import React, {
  cloneElement,
  FC,
  forwardRef,
  useEffect,
  useRef,
  useState,
} from 'react';
import {
  autoUpdate,
  ElementProps,
  flip,
  FloatingFocusManager,
  FloatingNode,
  FloatingPortal,
  FloatingTree,
  FloatingTreeType,
  offset,
  ReferenceType,
  safePolygon,
  shift,
  useClick,
  useDismiss,
  useFloating,
  useFloatingNodeId,
  useFloatingParentNodeId,
  useFloatingTree,
  useHover,
  useInteractions,
  useListNavigation,
  useMergeRefs,
  useRole,
} from '@floating-ui/react';
import { DropdownMenuProps, MenuItemTypes, MenuSize } from './Menu.types';
import { MenuItemType } from './MenuItem/MenuItem.types';
import { MenuItem } from './MenuItem/MenuItem';
import { Button, ButtonSize, ButtonVariant } from '../Button';
import { List } from '../List';
import { Stack } from '../Stack';
import { useCanvasDirection } from '../../hooks/useCanvasDirection';
import {
  canUseDom,
  eventKeys,
  mergeClasses,
  RenderProps,
} from '../../shared/utilities';

import dropdownStyles from '../Dropdown/dropdown.module.scss';
import menuStyles from './menu.module.scss';

export const MenuComponent: FC<DropdownMenuProps> = forwardRef<
  HTMLDivElement,
  DropdownMenuProps
>(({ children, ...props }, ref: React.ForwardedRef<HTMLDivElement>) => {
  const htmlDir: string = useCanvasDirection();

  const [isOpen, setIsOpen] = useState<boolean>(false);
  const [activeIndex, setActiveIndex] = useState<number | null>(null);
  const [allowHover, setAllowHover] = useState<boolean>(false);
  const tree: FloatingTreeType<ReferenceType> = useFloatingTree();
  const nodeId: string = useFloatingNodeId();
  const parentId: string = useFloatingParentNodeId();
  const isNested: boolean = parentId != null;

  const { x, y, strategy, refs, context } = useFloating<ReferenceType>({
    nodeId,
    open: isOpen,
    onOpenChange: setIsOpen,
    placement: isNested ? 'right-start' : 'bottom-start',
    middleware: [
      offset({
        mainAxis: isNested ? 8 : 0,
        alignmentAxis: isNested ? -8 : 0,
      }),
      flip(),
      shift(),
    ],
    whileElementsMounted: autoUpdate,
  });

  const listRef: React.MutableRefObject<any[]> = useRef([]);

  const listNavigation: ElementProps = useListNavigation(context, {
    listRef,
    activeIndex: activeIndex,
    nested: isNested,
    onNavigate: setActiveIndex,
    scrollItemIntoView: true,
  });

  const hover: ElementProps = useHover<ReferenceType>(context, {
    enabled: isNested && allowHover,
    delay: { open: 75 },
    handleClose: safePolygon({
      restMs: 25,
      blockPointerEvents: true,
    }),
  });

  const click: ElementProps = useClick<ReferenceType>(context, {
    event: 'mousedown',
    toggle: !isNested || !allowHover,
    ignoreMouse: isNested,
  });

  const role: ElementProps = useRole<ReferenceType>(context, {
    role: 'menu',
  });

  const dismiss: ElementProps = useDismiss<ReferenceType>(context);

  // TODO: ENG-46500 Implement `typeahead` in `useInteractions` via
  // `useTypeahead` floating-ui helper.
  // See floating-ui API reference implementation at:
  // https://codesandbox.io/s/bold-panna-9tf226?file=/src/DropdownMenu.tsx
  const { getReferenceProps, getFloatingProps, getItemProps } = useInteractions(
    [hover, click, role, dismiss, listNavigation]
  );

  // Event emitter allows you to communicate across tree components.
  // This effect closes all menus when an item gets clicked anywhere
  // in the tree.
  useEffect(() => {
    if (!tree) return null;

    const handleTreeClick = (): void => {
      setIsOpen(false);
    };

    const onSubMenuOpen = (event: {
      nodeId: string;
      parentId: string;
    }): void => {
      if (event.nodeId !== nodeId && event.parentId === parentId) {
        setIsOpen(false);
      }
    };

    tree.events.on('click', handleTreeClick);
    tree.events.on('menuopen', onSubMenuOpen);

    return () => {
      tree.events.off('click', handleTreeClick);
      tree.events.off('menuopen', onSubMenuOpen);
    };
  }, [tree, nodeId, parentId]);

  useEffect(() => {
    if (isOpen && tree) {
      tree.events.emit('menuopen', { parentId, nodeId });
    }
  }, [tree, isOpen, nodeId, parentId]);

  // Determine if "hover" logic can run based on the modality of input. This
  // prevents unwanted focus synchronization as menus open and close with
  // keyboard navigation and the cursor is resting on the menu.
  useEffect(() => {
    const onPointerMove = ({ pointerType }: PointerEvent): void => {
      if (pointerType !== 'touch') {
        setAllowHover(true);
      }
    };

    const onKeyDown = (): void => {
      setAllowHover(false);
    };

    if (canUseDom()) {
      window.addEventListener('pointermove', onPointerMove, {
        once: true,
        capture: true,
      });

      window.addEventListener('keydown', onKeyDown, true);
    }

    return () => {
      if (canUseDom()) {
        window.removeEventListener('pointermove', onPointerMove, {
          capture: true,
        });
        window.removeEventListener('keydown', onKeyDown, true);
      }
    };
  }, [allowHover]);

  const reference: (instance: ReferenceType) => void =
    useMergeRefs<ReferenceType>([refs.setReference, ref]);

  const handleReferenceKeyDown = (event: React.KeyboardEvent): void => {
    event?.stopPropagation();
    if (props.disabled) {
      return;
    }
    if (!isNested) {
      props.referenceOnKeydown?.(event);
    }
    if (
      event?.key === eventKeys.ARROWDOWN &&
      document.activeElement === event.target &&
      !isNested &&
      !isOpen
    ) {
      event?.preventDefault();
      setIsOpen(true);
    }
    if (
      event?.key === eventKeys.ARROWUP &&
      document.activeElement === event.target &&
      !isNested &&
      isOpen
    ) {
      event?.preventDefault();
      setIsOpen(false);
    }
    if (
      event?.key === eventKeys.ESCAPE ||
      (event?.key === eventKeys.TAB &&
        event.shiftKey &&
        !(event.target as HTMLElement).matches(':focus-within'))
    ) {
      setIsOpen(false);
    }
  };

  const getReference = (): JSX.Element | React.ReactNode => {
    if (React.isValidElement(children)) {
      const child = React.Children.only(children) as React.ReactElement<any>;
      const referenceClassNames: string = mergeClasses([
        // Add any classnames added to the reference element
        { [child.props.className]: !!child.props.className },
        { [child.props.classNames]: !!child.props.classNames },
      ]);

      const clonedElementProps: RenderProps = {
        ref: reference,
        role: 'button',
        'data-open': isOpen ? '' : undefined,
        ...getReferenceProps({
          className: referenceClassNames,
          onClick(event) {
            event.stopPropagation();
            if (!isNested) {
              props.referenceOnClick?.(event);
            }
          },
          onKeyDown: handleReferenceKeyDown,
          ...(isNested && {
            role: 'menuitem',
          }),
        }),
      };

      // Compares for octuple react prop vs native react html classes.
      if (child.props?.className) {
        clonedElementProps['className'] = referenceClassNames;
      } else if (child.props.classNames) {
        clonedElementProps['classNames'] = referenceClassNames;
      }

      return cloneElement(child, clonedElementProps);
    }
    return children;
  };

  const footerClassNames: string = mergeClasses([
    menuStyles.menuFooterContainer,
    {
      [menuStyles.large]: props.size === MenuSize.large,
      [menuStyles.medium]: props.size === MenuSize.medium,
      [menuStyles.small]: props.size === MenuSize.small,
    },
  ]);

  const headerClassNames: string = mergeClasses([
    menuStyles.menuHeaderContainer,
    {
      [menuStyles.large]: props.size === MenuSize.large,
      [menuStyles.medium]: props.size === MenuSize.medium,
      [menuStyles.small]: props.size === MenuSize.small,
    },
  ]);

  const menuClassNames: string = mergeClasses([
    props.classNames,
    menuStyles.menuContainer,
    {
      [menuStyles.large]: props.size === MenuSize.large,
      [menuStyles.medium]: props.size === MenuSize.medium,
      [menuStyles.small]: props.size === MenuSize.small,
    },
  ]);

  const menuSizeToButtonSizeMap: Map<MenuSize, ButtonSize> = new Map<
    MenuSize,
    ButtonSize
  >([
    [MenuSize.large, ButtonSize.Large],
    [MenuSize.medium, ButtonSize.Medium],
    [MenuSize.small, ButtonSize.Small],
  ]);

  const getListItem = (item: MenuItemTypes, index: number): JSX.Element => (
    <MenuItem
      classNames={props.itemClassNames}
      direction={htmlDir}
      key={`oc-menu-item-${index}`}
      onChange={props.onChange}
      onClick={(event: React.MouseEvent<HTMLDivElement, MouseEvent>) => {
        props.onClick?.(event);
        tree?.events.emit('click');
        props.onChange?.(item.value);
      }}
      onMouseEnter={() => {
        if (allowHover && isOpen) {
          setActiveIndex(index);
        }
      }}
      ref={(item: HTMLElement) => {
        listRef.current[index] = item;
      }}
      size={props.size}
      tabIndex={activeIndex === index && 0}
      type={item.type ?? MenuItemType.button}
      variant={props.variant}
      {...item}
      {...getItemProps()}
    />
  );

  const getHeader = (): JSX.Element =>
    props.header && (
      <div className={headerClassNames}>
        <div className={menuStyles.heading}>{props.header}</div>
      </div>
    );

  const getFooter = (): JSX.Element =>
    (props.cancelButtonProps || props.okButtonProps) && (
      <Stack
        flexGap="s"
        justify="flex-end"
        fullWidth
        classNames={footerClassNames}
      >
        {props.cancelButtonProps && (
          <Button
            variant={ButtonVariant.Neutral}
            {...props.cancelButtonProps}
            size={menuSizeToButtonSizeMap.get(props.size)}
            onClick={props.onCancel}
          />
        )}
        {props.okButtonProps && (
          <Button
            variant={ButtonVariant.Primary}
            {...props.okButtonProps}
            size={menuSizeToButtonSizeMap.get(props.size)}
            onClick={props.onOk}
          />
        )}
      </Stack>
    );

  return (
    <FloatingNode id={nodeId}>
      {getReference()}
      <FloatingPortal>
        {isOpen && (
          <FloatingFocusManager
            context={context}
            // Prevent outside content interference.
            modal={false}
            initialFocus={0}
            // Only return focus to the root menu's reference when menus close.
            returnFocus={!isNested}
          >
            <div
              ref={refs.setFloating}
              className={mergeClasses([
                dropdownStyles.dropdownWrapper,
                dropdownStyles.noPadding,
                { [dropdownStyles.open]: isOpen },
                { [dropdownStyles.close]: !isOpen },
              ])}
              style={{
                position: strategy,
                top: y ?? 0,
                left: x ?? 0,
                width: 'max-content',
              }}
              {...getFloatingProps()}
            >
              <List<MenuItemTypes>
                {...props}
                classNames={menuClassNames}
                footer={getFooter()}
                getItem={getListItem}
                header={getHeader()}
                items={props.items}
                listType={props.listType}
                role="menu"
                style={props.style}
              />
            </div>
          </FloatingFocusManager>
        )}
      </FloatingPortal>
    </FloatingNode>
  );
});

export const CascadingMenu: FC<DropdownMenuProps> = forwardRef<
  HTMLDivElement,
  DropdownMenuProps
>(
  (
    props: React.PropsWithChildren<DropdownMenuProps>,
    ref: React.ForwardedRef<HTMLDivElement>
  ) => {
    const parentId: string = useFloatingParentNodeId();

    if (parentId === null) {
      return (
        <FloatingTree>
          <MenuComponent {...props} ref={ref} />
        </FloatingTree>
      );
    }

    return <MenuComponent {...props} ref={ref} />;
  }
);
