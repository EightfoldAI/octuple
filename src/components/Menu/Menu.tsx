'use client';

import React, { FC } from 'react';
import { MenuItemTypes, MenuProps, MenuSize, MenuVariant } from './Menu.types';
import { List } from '../List';
import { MenuItem } from './MenuItem/MenuItem';
import { MenuItemType } from './MenuItem/MenuItem.types';
import { Stack } from '../Stack';
import { Button, ButtonSize, ButtonVariant } from '../Button';
import { useCanvasDirection } from '../../hooks/useCanvasDirection';
import { mergeClasses } from '../../shared/utilities';

import styles from './menu.module.scss';

export const Menu: FC<MenuProps> = ({
  cancelButtonProps,
  classNames,
  header,
  itemClassNames,
  itemProps,
  items,
  itemStyle,
  listType,
  okButtonProps,
  onCancel,
  onChange,
  onOk,
  role = 'menu',
  size = MenuSize.medium,
  style,
  subHeader, // TODO: Remove in v3.0.0
  variant = MenuVariant.neutral,
  ...rest
}) => {
  const htmlDir: string = useCanvasDirection();

  const footerClassNames: string = mergeClasses([
    styles.menuFooterContainer,
    {
      [styles.large]: size === MenuSize.large,
      [styles.medium]: size === MenuSize.medium,
      [styles.small]: size === MenuSize.small,
    },
  ]);

  const headerClassNames: string = mergeClasses([
    styles.menuHeaderContainer,
    {
      [styles.large]: size === MenuSize.large,
      [styles.medium]: size === MenuSize.medium,
      [styles.small]: size === MenuSize.small,
    },
  ]);

  const menuClassNames: string = mergeClasses([
    classNames,
    styles.menuContainer,
    {
      [styles.large]: size === MenuSize.large,
      [styles.medium]: size === MenuSize.medium,
      [styles.small]: size === MenuSize.small,
    },
  ]);

  const menuSizeToButtonSizeMap: Map<MenuSize, ButtonSize> = new Map<
    MenuSize,
    ButtonSize
  >([
    [MenuSize.large, ButtonSize.Large],
    [MenuSize.medium, ButtonSize.Medium],
    [MenuSize.small, ButtonSize.Small],
  ]);

  const getListItem = (item: MenuItemTypes, index: number): React.ReactNode => (
    <MenuItem
      classNames={itemClassNames}
      direction={htmlDir}
      key={`oc-menu-item-${index}`}
      variant={variant}
      size={size}
      onClick={onChange}
      onChange={onChange}
      type={item.type ?? MenuItemType.button}
      {...item}
    />
  );

  const getHeader = (): JSX.Element =>
    header && (
      <div className={headerClassNames}>
        <div className={styles.heading}>{header}</div>
      </div>
    );

  const getFooter = (): JSX.Element =>
    (cancelButtonProps || okButtonProps) && (
      <Stack
        flexGap="s"
        justify="flex-end"
        fullWidth
        classNames={footerClassNames}
      >
        {cancelButtonProps && (
          <Button
            variant={ButtonVariant.Neutral}
            {...cancelButtonProps}
            size={menuSizeToButtonSizeMap.get(size)}
            onClick={onCancel}
          />
        )}
        {okButtonProps && (
          <Button
            variant={ButtonVariant.Primary}
            {...okButtonProps}
            size={menuSizeToButtonSizeMap.get(size)}
            onClick={onOk}
          />
        )}
      </Stack>
    );

  return (
    <List<MenuItemTypes>
      {...rest}
      classNames={menuClassNames}
      footer={getFooter()}
      getItem={getListItem}
      header={getHeader()}
      items={items}
      listType={listType}
      role={role}
      style={style}
    />
  );
};
