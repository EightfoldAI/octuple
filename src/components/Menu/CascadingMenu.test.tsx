import React from 'react';
import Enzyme from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import MatchMediaMock from 'jest-matchmedia-mock';
import { MenuItemIconAlign, MenuItemType } from './MenuItem/MenuItem.types';
import { MenuVariant } from './Menu.types';
import { CascadingMenu } from './CascadingMenu';
import { Button } from '../Button';
import { IconName } from '../Icon';
import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom';

Enzyme.configure({ adapter: new Adapter() });

let matchMedia: any;

class ResizeObserver {
  observe() {
    // do nothing
  }
  unobserve() {
    // do nothing
  }
  disconnect() {
    // do nothing
  }
}

window.ResizeObserver = ResizeObserver;

const cascadingMenuItems = [
  {
    iconProps: {
      path: IconName.mdiCalendar,
    },
    text: 'Button',
    value: 'menu 1',
    counter: '8',
    'data-testid': 'firstMenuElement',
    secondaryButtonProps: {
      iconProps: {
        path: IconName.mdiTrashCan,
      },
    },
  },
  {
    text: 'Disabled button',
    value: 'menu 2',
    disabled: true,
    subText: 'This is a sub text',
  },
  {
    iconProps: {
      path: IconName.mdiCalendar,
    },
    text: 'Date',
    value: 'menu 3',
    counter: '8',
  },
  {
    alignIcon: MenuItemIconAlign.Right,
    iconProps: {
      path: IconName.mdiChevronRight,
    },
    dropdownMenuItems: [
      {
        text: 'Button 1',
        value: 'subMenuA 1',
      },
      {
        text: 'Button 2',
        value: 'subMenuA 1',
      },
      {
        text: 'Button 3',
        value: 'subMenuA 1',
      },
      {
        alignIcon: MenuItemIconAlign.Right,
        iconProps: {
          path: IconName.mdiChevronRight,
        },
        dropdownMenuItems: [
          {
            text: 'Button 4',
            value: 'subMenuB 1',
          },
          {
            text: 'Button 5',
            value: 'subMenuB 2',
          },
          {
            text: 'Button 6',
            value: 'subMenuB 3',
          },
        ],
        text: 'Sub menu 2',
        value: 'subMenuA 2',
      },
    ],
    text: 'Sub menu 1',
    value: 'menu 4',
  },
  {
    type: MenuItemType.subHeader,
    text: 'Sub header',
  },
  {
    type: MenuItemType.link,
    text: 'Twitter link',
    href: 'https://twitter.com',
    target: '_blank',
  },
];

const menuProps: Object = {
  variant: MenuVariant.neutral,
  classNames: 'my-menu-class',
  style: {
    color: 'red',
  },
  itemClassNames: 'my-menu-item-class',
  itemStyle: {},
  listType: 'ul',
};

const CascadingMenuComponent = (): JSX.Element => (
  <CascadingMenu {...menuProps} items={cascadingMenuItems}>
    <Button data-testid="menu-reference" text={'Cascading menu'} />
  </CascadingMenu>
);

describe('Menu', () => {
  beforeAll(() => {
    matchMedia = new MatchMediaMock();
  });

  afterEach(() => {
    matchMedia.clear();
  });

  test('Should render a menu button', () => {
    const { container } = render(<CascadingMenuComponent />);
    const dropdownButton = screen.getByRole('button');
    expect(dropdownButton).toBeTruthy();
    expect(() => container).not.toThrowError();
    expect(container).toMatchSnapshot();
  });

  test('Should show the menu items when the button is clicked', async () => {
    render(<CascadingMenuComponent />);
    const dropdownButton = screen.getByRole('button');
    dropdownButton.click();
    await waitFor(() => screen.getByText('Button'));
    const menuitem1 = screen.getByText('Button');
    const menuitem2 = screen.getByText('Disabled button');
    const menuitem3 = screen.getByText('Date');
    const menuitem4 = screen.getByText('Sub menu 1');
    const menuitem5 = screen.getByText('Twitter link');
    expect(menuitem1).toBeTruthy();
    expect(menuitem2).toBeTruthy();
    expect(menuitem3).toBeTruthy();
    expect(menuitem4).toBeTruthy();
    expect(menuitem5).toBeTruthy();
  });

  test('Should show the sub menu items when the button is clicked', async () => {
    render(<CascadingMenuComponent />);
    const dropdownButton = screen.getByRole('button');
    dropdownButton.click();
    await waitFor(() => screen.getByText('Sub menu 1'));
    const submenuitem = screen.getByText('Sub menu 1');
    submenuitem.click();
    await waitFor(() => screen.getByText('Button 1'));
    const menuitem1 = screen.getByText('Button 1');
    const menuitem2 = screen.getByText('Button 2');
    const menuitem3 = screen.getByText('Button 3');
    const menuitem4 = screen.getByText('Sub menu 2');
    expect(menuitem1).toBeTruthy();
    expect(menuitem2).toBeTruthy();
    expect(menuitem3).toBeTruthy();
    expect(menuitem4).toBeTruthy();
  });

  test('Should show the second sub menu items when the first sub menu button is clicked', async () => {
    render(<CascadingMenuComponent />);
    const dropdownButton = screen.getByRole('button');
    dropdownButton.click();
    await waitFor(() => screen.getByText('Sub menu 1'));
    const submenuitem1 = screen.getByText('Sub menu 1');
    submenuitem1.click();
    await waitFor(() => screen.getByText('Sub menu 2'));
    const submenuitem2 = screen.getByText('Sub menu 2');
    submenuitem2.click();
    await waitFor(() => screen.getByText('Button 4'));
    const menuitem1 = screen.getByText('Button 4');
    const menuitem2 = screen.getByText('Button 5');
    const menuitem3 = screen.getByText('Button 6');
    expect(menuitem1).toBeTruthy();
    expect(menuitem2).toBeTruthy();
    expect(menuitem3).toBeTruthy();
  });

  test('Calls the referenceOnClick function when a key is pressed', () => {
    const referenceOnClick = jest.fn();
    const { getByTestId } = render(
      <CascadingMenu
        {...menuProps}
        items={cascadingMenuItems}
        referenceOnClick={referenceOnClick}
      >
        <Button data-testid="menu-reference" text={'Cascading menu'} />
      </CascadingMenu>
    );
    const referenceElement = getByTestId('menu-reference');
    fireEvent.click(referenceElement);
    expect(referenceOnClick).toHaveBeenCalledTimes(1);
  });

  test('Calls the referenceOnKeydown function when a key is pressed', () => {
    const referenceOnKeydown = jest.fn();
    const { getByTestId } = render(
      <CascadingMenu
        {...menuProps}
        items={cascadingMenuItems}
        referenceOnKeydown={referenceOnKeydown}
      >
        <Button data-testid="menu-reference" text={'Cascading menu'} />
      </CascadingMenu>
    );
    const referenceElement = getByTestId('menu-reference');
    fireEvent.keyDown(referenceElement, { key: 'ArrowDown' });
    expect(referenceOnKeydown).toHaveBeenCalledTimes(1);
  });

  test('Arrow down key opens the menu and places focus on the first item', async () => {
    render(<CascadingMenuComponent />);
    const referenceButton = screen.getByTestId('menu-reference');
    referenceButton.focus();
    fireEvent.keyDown(referenceButton, { key: 'ArrowDown' });
    const firstMenuItem = screen.getByTestId('firstMenuElement');
    await waitFor(() => expect(firstMenuItem).toBeInTheDocument());
    expect(firstMenuItem).toHaveFocus();
  });

  test('Arrow up key closes the menu and places focus back on the reference element', async () => {
    render(<CascadingMenuComponent />);
    const referenceButton = screen.getByTestId('menu-reference');
    fireEvent.keyDown(referenceButton, { key: 'ArrowDown' });
    const menu = screen.getAllByRole('menu')[0];
    expect(menu).toBeInTheDocument();
    referenceButton.focus();
    fireEvent.keyDown(referenceButton, { key: 'ArrowUp' });
    await waitFor(() => expect(menu).not.toBeInTheDocument());
    expect(referenceButton).toHaveFocus();
  });

  test('Escape key closes the menu and places focus back on the reference element', async () => {
    render(<CascadingMenuComponent />);
    const referenceButton = screen.getByTestId('menu-reference');
    fireEvent.keyDown(referenceButton, { key: 'ArrowDown' });
    const menu = screen.getAllByRole('menu')[0];
    expect(menu).toBeInTheDocument();
    const firstMenuItem = screen.getByText('Button');
    fireEvent.keyDown(firstMenuItem, { key: 'Escape' });
    await waitFor(() => expect(menu).not.toBeInTheDocument());
    expect(referenceButton).toHaveFocus();
  });
});
