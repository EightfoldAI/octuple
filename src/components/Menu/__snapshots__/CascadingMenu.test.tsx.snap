// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu Should render a menu button 1`] = `
<div>
  <button
    aria-disabled="false"
    aria-expanded="false"
    aria-haspopup="menu"
    class="button button-default button-medium pill-shape"
    data-testid="menu-reference"
    id="floating-ui-3"
    role="button"
  >
    <span
      class="button-text-medium"
    >
      Cascading menu
    </span>
  </button>
</div>
`;
